File: php/src/FnDsa/Verify.php
<?php

declare(strict_types=1);

namespace PQC\FnDsa;

/**
 * FN-DSA verification (FIPS 206 Algorithm 4).
 */
final class Verify
{
    private const Q = 12289;

    /**
     * Verify a FN-DSA signature.
     *
     * @param string $pk Encoded public key.
     * @param string $msg Message bytes.
     * @param string $sig Encoded signature.
     * @param Params $p Parameter set.
     * @return bool True if valid.
     */
    public static function verify(string $pk, string $msg, string $sig, Params $p): bool
    {
        // 1. Decode and validate public key.
        $h = Encode::decodePk($pk, $p);
        if ($h === null) {
            return false;
        }

        // 2. Decode and validate signature.
        $decoded = Encode::decodeSig($sig, $p);
        if ($decoded === null) {
            return false;
        }
        [$salt, $s1] = $decoded;

        // 3. Recompute c = HashToPoint(salt || msg).
        $hashInput = $salt . $msg;
        $c = Sign::hashToPoint($hashInput, $p);

        // 4. Compute s2 = c - s1*h (mod q), centered.
        $n = $p->n;
        $s1ModQ = [];
        for ($i = 0; $i < $n; $i++) {
            $s1ModQ[$i] = (($s1[$i] % self::Q) + self::Q) % self::Q;
        }
        $s1h = NTT::polyMulNtt($s1ModQ, $h, $n);
        $s2 = [];
        for ($i = 0; $i < $n; $i++) {
            $s2[$i] = Sign::centerModQ($c[$i] - $s1h[$i]);
        }

        // 5. Norm check: ||(s1, s2)||^2 <= beta^2.
        return Sign::normSq($s1, $s2) <= $p->betaSq;
    }
}

File: python/fndsa/verify.py
"""FN-DSA verification (FIPS 206 Algorithm 4)."""

from .params import Params, Q
from .encode import decode_pk, decode_sig
from .ntt import poly_mul_ntt
from .sign import hash_to_point


def _center_mod_q(v: int) -> int:
    """Reduce v mod Q and center in (-Q/2, Q/2]."""
    v = ((v % Q) + Q) % Q
    if v > Q // 2:
        v -= Q
    return v


def _norm_sq(s1: list[int], s2: list[int]) -> int:
    """Compute squared Euclidean norm of (s1, s2)."""
    return sum(v * v for v in s1) + sum(v * v for v in s2)


def verify(pk: bytes, msg: bytes, sig: bytes, params: Params) -> bool:
    """Verify an FN-DSA signature (FIPS 206 Algorithm 4).

    Returns True iff sig is a valid FN-DSA signature on msg under public key pk
    for parameter set params.
    """
    # 1. Decode and validate public key
    h = decode_pk(pk, params)
    if h is None:
        return False

    # 2. Decode and validate signature
    result = decode_sig(sig, params)
    if result is None:
        return False
    salt, s1 = result

    # 3. Recompute c = HashToPoint(salt || msg)
    hash_input = salt + msg
    c = hash_to_point(hash_input, params)

    # 4. Compute s2 = c - s1*h (mod q), centered in (-Q/2, Q/2]
    n = params.n
    s1_mod_q = [((v % Q) + Q) % Q for v in s1]
    s1h = poly_mul_ntt(s1_mod_q, h, n)
    s2 = [_center_mod_q(c[i] - s1h[i]) for i in range(n)]

    # 5. Norm check: ||(s1, s2)||^2 <= beta^2
    return _norm_sq(s1, s2) <= params.beta_sq
